### PR TITLE
Add in some meta targets

### PR DESCRIPTION
**Pull request by ianoc-stripe**
Both test and library meta targets

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

File: scala/scala.bzl [hunk 1]
```diff
@@ -656,3 +656,52 @@ def scala_export_to_java(name, exports, runtime_deps):
...
     runtime_deps = ["@scala//:lib/scala-library.jar"] + runtime_deps
   )
+
+# This auto-generates a test suite based on the passed set of targets
+# we will add a root test_suite with the name of the passed name
+def scala_test_suite(name, srcs = [], deps = [], runtime_deps = [], data = [], resources = [],
+                     scalacopts = [], jvm_flags = [], visibility = None, size = None):
+    ts = []
+    for test_file in srcs:
+        n = "suite_test_%s_%s_test" % (name, test_file)
```
Comment (johnynek) [inline scala/scala.bzl]: do we need to normalize the test_file names at all? Remove spaces, lower case?

File: scala/scala.bzl [hunk 2]
```diff
@@ -656,3 +656,52 @@ def scala_export_to_java(name, exports, runtime_deps):
...
+                        resource_strip_prefix = "",
+                        scalacopts = [],
+                        javacopts = [],
+                        jvm_flags = [],
+                        print_compile_time = False,
+                        visibility = None
+                        ):
+    ts = []
+    for src_file in srcs:
+        n = "%s_lib_%s" % (name, src_file)
```
Comment (johnynek) [inline scala/scala.bzl]: same comment/question about `src_file`.

File: scala/scala.bzl [hunk 3]
```diff
@@ -656,3 +656,52 @@ def scala_export_to_java(name, exports, runtime_deps):
...
+                      plugins = plugins,
+                      runtime_deps = runtime_deps,
+                      data = data,
+                      resources=resources,
+                      resource_strip_prefix = resource_strip_prefix,
+                      scalacopts = scalacopts,
+                      javacopts = javacopts,
+                      jvm_flags = jvm_flags,
+                      print_compile_time = print_compile_time,
+                      visibility=visibility,
```
Comment (johnynek) [inline scala/scala.bzl]: should we expose the inner targets, or should we make them private, and only use the visibility on the item below?

File: scala/scala.bzl [hunk 4]
```diff
@@ -656,3 +656,52 @@ def scala_export_to_java(name, exports, runtime_deps):
...
+                      exports=exports
+                      )
+        ts.append(n)
+    scala_library(name = name, exports = exports + ts)
```
Comment (johnynek) [inline scala/scala.bzl]: don't we need to set visibility here?

File: test/BUILD [hunk 5]
```diff
@@ -6,7 +6,8 @@ load("//scala:scala.bzl",
     "scala_test",
     "scala_macro_library",
     "scala_export_to_java",
-    "scala_repl")
+    "scala_repl",
+    "scala_library_suite")
```
Comment (johnynek) [inline test/BUILD]: we don't yet have a test for `scala_test_suite` do we?

File: scala/scala.bzl (hunk 1 above)
Comment (ianoc-stripe) [inline scala/scala.bzl]: I'll add a sanitize, but cases it doesn't care about upper/lower case. And spaces aren't valid seemingly in input file names to scala library at all.

File: scala/scala.bzl (hunk 3 above)
Comment (ianoc-stripe) [inline scala/scala.bzl]: Thats a good question, i'd have to go with i don't know. You could depend on the inner targets this way, and i know Avi expressed some interest in being able to do that. But then the generated target names need to be stable for referring in build files.

File: test/BUILD (hunk 5 above)
Comment (ianoc-stripe) [inline test/BUILD]: I had added some of the code below, forgot to wire it in the final bit.

**Comment by ianoc-stripe** [comment]
Updated the readme, only outstanding thing is the visibility. I'm pretty good with either answer there.

File: scala/scala.bzl (hunk 3 above)
Comment (johnynek) [inline scala/scala.bzl]: 👍 I guess we can just stay as is now and see later. Too public is probably better than too private for now.

**Comment by johnynek** [comment]
@ittaiz it will not work if there are *ANY* dependencies between the files. They must be *independent* (which it does mention in the readme).

**Comment by ittaiz** [comment]
@johnynek indeed it states that it will split the independent files. It wasn't clear to me if this can be of value if it can handle dependencies. I now understand this is more of a naive parallelization.
Is this really valuable in the library level? I can understand that tests are likely to be decoupled by nature but 10 files in the same source files usually have some relationship between them, don't they?

**Comment by johnynek** [comment]
I think seldom will it help for libraries but frequently help with tests.

**Comment by ittaiz** [comment]
Got you, that's what I thought. Thanks!

**Pull request merged**